finality-aleph/src/network.rs

```diff
@@ -404,11 +403,21 @@ impl<D: Clone + Encode + Decode, B: BlockT + 'static, N: Network<B> + Clone>
         }
     }
 
-    fn on_incoming_data(&self, session_id: SessionId, data: D) {
+    async fn on_incoming_data(&self, session_id: SessionId, data: NetworkData<RushData, RmcData>) {
         let mut sessions = self.sessions.lock();
         if let Some(session_data) = sessions.get_mut(&session_id) {
             if session_data.status == SessionStatus::InProgress {
-                if let Err(e) = session_data.data_for_user.unbounded_send(data) {
+                let result = match data {
```

Comment:
Uh, major complaint incoming, sorry.

This is the only place in all of `ConsensusNetwork` where `D = NetworkData<RushData, RmcData>` actually matters. All the rest of the logic operates on a completely arbitrary `D: Clone + Codec`. I think this is not a coincidence, and suggests the `NetworkData`-specific logic should be in a separate entity (I'll call it hub below for clarity, but feel free to come up with a better name), while this remains a very generic network interface for arbitrary data.

The main disadvantage would be that this hub has to be run as another (lite)thread, to actually do the logic  here, i.e. unpacking the incoming messages and separating them into different channels. I don't think this problematic, and readability would definitely improve.

This change would also coalesce `RushNetwork` and `RmcNetwork` into one type, again parametrized by the type of data being sent. The packing into `NetworkData::Rush/Rmc` would only happen in the hub and it would be the only struct aware of the internals of `NetworkData` or even the fact that there is more than one kind of network messages at all.

Reply:
As a bonus all this would be easier to unit test, component by component.

Reply:
Agree, will change that